src/main/java/core/basesyntax/MyHashMap.java

```diff
@@ -1,19 +1,117 @@
 package core.basesyntax;
 
+import java.util.Objects;
+
 public class MyHashMap<K, V> implements MyMap<K, V> {
+    private static final int INITIAL_CAPACITY = 16;
+    private static final float LOAD_FACTOR = 0.75f;
+    private static final int GROW_NUMBER = 2;
+    private Node<K, V>[] table;
+    private int size;
+
+    public MyHashMap() {
+        table = (Node<K, V>[]) new Node[INITIAL_CAPACITY];
+    }
 
     @Override
     public void put(K key, V value) {
+        ensureCapacity();
+
+        Node<K, V> node = new Node<>(key, value, null);
+        int index = getKeyBucket(key);
+        if (table[index] == null) {
+            table[index] = node;
+            size++;
+            return;
+        }
+        Node<K, V> currentNode = table[index];
+        while (currentNode != null) {
+            if (Objects.equals(currentNode.key, key)) {
+                currentNode.value = value;
+                return;
+            }
+            if (currentNode.next == null) {
+                currentNode.next = node;
+                size++;
+            }
+            currentNode = currentNode.next;
+        }
 
     }
 
     @Override
     public V getValue(K key) {
+        Node<K, V> node = table[getKeyBucket(key)];
+        while (node != null) {
+            if (Objects.equals(node.key, key)) {
+                return node.value;
+            }
+            node = node.next;
+        }
         return null;
     }
 
     @Override
     public int getSize() {
-        return 0;
+        return size;
+    }
+
+    private int hash(K key) {
+        int h;
+        return (key == null) ? 0 : Math.abs((h = key.hashCode()) ^ (h >>> 16));
```

Comment:
```suggestion
        int h = key.hashCode();
        return (key == null) ? 0 : Math.abs(h) ^ (h >>> 16));
```

Reply:
If i change this, then i get NPE. If i do this way, should i add some checking condition if( key == null)? but than code will be bigger i think?